Modules/rat_facts.py

```diff
@@ -0,0 +1,209 @@
+        checks for existence of a fact
+        Args:
+            name: name of the fact to check
+            lang: language to check
+
+        Returns: True if it exists, False if it does not exist
+
+        """
+        # this is a intended apparent duplicate of get_fact
+        result = await self.dbm.select_rows("fact", "AND", {"lang": lang, "name": name})
+        return len(result) > 0
+
+    async def set_fact(self, fact: Fact) -> bool:
+        """
+        sets a fact
+        Args:
+            fact (Fact): the fact to push into the DB
+        Returns: True on success, False if it failed
+
+        """
+        try:
+            await self.dbm.insert_row("fact",
+                                      {"name": fact.name,
+                                       "lang": fact.lang,
+                                       "message": fact.message,
+                                       "author": fact.author},
+                                      ("name", "lang",)
+                                      )
+        except ValueError:
+            return False
+        else:
+            time = datetime.datetime.now(tz=datetime.timezone.utc)
+            old_fact = await self.get_fact(fact.name, fact.lang)
+            if old_fact.message == fact.message:
+                self.dbm._execute(r"INSERT INTO fact_timestamps ("
+                                  r"name, lang, message, author, date_changed) "
+                                  r"VALUES("
+                                  f" '{fact.name}',"
+                                  f" '{fact.lang}',"
+                                  f" '{old_fact.message}',"
+                                  f" '{old_fact.author}', ?"
+                                  f")",
+                                  time
+                                  )
+
+            return True
+
+    async def delete_fact(self, name: str, lang: str) -> bool:
+        """
+        delete's the given fact
+        Args:
+            name: name of the fact to delete
+            lang: language to delete
+
+        Returns: True on success, False otherwise
+
+        """
+        try:
+            await self.dbm.delete_row("fact", "AND", {"name": name, "lang": lang})
+            return True
+        except ValueError:
+            return False
+
+    async def handle_fact(self, context: Context) -> bool:
+        """
+
+        Responds in IRC with the message of the fact, should a match be found and not listed as an
+        internal fact.
+
+        Args:
+            context: the context as passed by rat_command
+
+        Returns: True if it was a fact, False if not. Will also return False,
+            should the DBM be disabled.
+
+        """
+        if not self.dbm.enabled:
+            await context.reply("Sadly, Facts are currently not available. Please check back later.")
+            return False
+
+        regex = re.compile(r"!(?P<name>[a-zA-Z\d]+)-(?P<lang>[a-zA-Z]{2})|!(?P<name2>[a-zA-Z\d]+)")
```

Comment:
If for no other reason other than Me not knowing enough of ReGex to read that, could you please document this regex so us non-regex magicians can understand it? ((not blocking))

Reply:
```suggestion
        regex = re.compile(r"!(?P<name>[a-zA-Z\d]+)-(?P<lang>[a-zA-Z]{2})|!(?P<name2>[a-zA-Z\d]+)")
        # Tries to split the inputted word (e.g !test-tr) into a group called "name" and a group called "lang" should this fail, for example because no language has been supplied, the it'll store the name of the fact in a group called "name2"
```